debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/AbstractLogMinerStreamingChangeEventSource.java

```diff
@@ -0,0 +1,1071 @@
+            return logCollector.getLogsForOffsetScn(scn).stream()
+                    .anyMatch(log -> log.isScnInLogFileRange(scn) && log.isArchive());
+        }
+        catch (LogFileNotFoundException e) {
+            // It is safe to ignore this error.
+            // This identifies that the check should simply be re-evaluated after the pause.
+            return false;
+        }
+    }
+
+    /**
+     * Calculates the deviated end scn based on the scn range and deviation.
+     *
+     * @param lowerboundsScn the mining range's lower bounds
+     * @param upperboundsScn the mining range's upper bounds
+     * @param deviation the time deviation
+     * @return an optional that contains the deviated scn or empty if the operation should be performed again
+     */
+    private Optional<Scn> calculateDeviatedEndScn(Scn lowerboundsScn, Scn upperboundsScn, Duration deviation) {
+        if (connectorConfig.isArchiveLogOnlyMode()) {
+            // When archive-only mode is enabled, deviation should be ignored, even when enabled.
+            return Optional.of(upperboundsScn);
+        }
+
+        final Optional<Scn> calculatedDeviatedEndScn = getDeviatedMaxScn(upperboundsScn, deviation);
+        if (calculatedDeviatedEndScn.isEmpty() || calculatedDeviatedEndScn.get().isNull()) {
+            // This happens only if the deviation calculation is outside the flashback/undo area or an exception was thrown.
+            // In this case we have no choice but to use the upper bounds as a fallback.
+            LOGGER.warn("Mining session end SCN deviation calculation is outside undo space, using upperbounds {}. If this continues, " +
+                    "consider lowering the value of the '{}' configuration property.", upperboundsScn,
+                    OracleConnectorConfig.LOG_MINING_MAX_SCN_DEVIATION_MS.name());
+            return Optional.of(upperboundsScn);
+        }
+        else if (calculatedDeviatedEndScn.get().compareTo(lowerboundsScn) <= 0) {
+            // This should also force the outer loop to recall this method again.
+            LOGGER.debug("Mining session end SCN deviation as {}, outside of mining range, recalculating.", calculatedDeviatedEndScn.get());
+            return Optional.empty();
+        }
+        else {
+            // Calculated SCN is after lower bounds and within flashback/undo area, safe to return.
+            return calculatedDeviatedEndScn;
+        }
+    }
+
+    /**
+     * Uses the provided Upperbound SCN and deviation to calculate an SCN that happened in the past at a
+     * time based on Oracle's {@code TIMESTAMP_TO_SCN} and {@code SCN_TO_TIMESTAMP} functions.
+     *
+     * @param upperboundsScn the upper bound system change number, should not be {@code null}
+     * @param deviation the time deviation to be applied, should not be {@code null}
+     * @return the newly calculated Scn
+     */
+    private Optional<Scn> getDeviatedMaxScn(Scn upperboundsScn, Duration deviation) {
```

Comment:
Something to consider.  

Not sure how often this method is called or how expensive the calls to the db are, but seems like we could possibly reduce the number of calls to the database in this method.

Reply:
It's called once per mining iteration and only if `internal.log.mining.max.scn.deviation.ms` is set, so I don't think it's really a concern with it being an internal, undocumented code path.